readonly_bs.go

```diff
@@ -0,0 +1,207 @@
+package dagstore
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/filecoin-project/dagstore/shard"
+	lru "github.com/hashicorp/golang-lru"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+)
+
+var logbs = logging.Logger("dagstore-all-readblockstore")
+
+var _ blockstore.Blockstore = (*AllShardsReadBlockstore)(nil)
+
+// ShardSelectorF helps select a shard to fetch a cid from if the given cid is present in multiple shards.
+type ShardSelectorF func(c cid.Cid, shards []shard.Key) (shard.Key, error)
+
+type accessorWithBlockstore struct {
+	sa *ShardAccessor
+	bs ReadBlockstore
+}
+
+// AllShardsReadBlockstore is a read only blockstore over all cids across all shards in the dagstore.
+type AllShardsReadBlockstore struct {
+	d            *DAGStore
+	shardSelectF ShardSelectorF
+
+	// caches the blockstore for a given shard for shard read affinity i.e. further reads will likely be from the same shard.
+	// shard key -> read only blockstore
+	blockstoreCache *lru.Cache
+
+	// caches the blocks themselves -> can be scaled by using a redis/memcache etc distributed cache.
+	// multihash -> block
+	blockCache *lru.Cache
+}
+
+func (d *DAGStore) AllShardsReadBlockstore(shardSelector ShardSelectorF, maxCacheSize int, maxBlocks int) (blockstore.Blockstore, error) {
+	// instantiate the blockstore cache
+	bslru, err := lru.NewWithEvict(maxCacheSize, func(_ interface{}, val interface{}) {
+		// ensure we close the blockstore for a shard when it's evicted from the cache so dagstore can gc it.
+		abs := val.(*accessorWithBlockstore)
+		abs.sa.Close()
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create lru cache for read only blockstores")
+	}
+
+	// instantiate the block cache
+	blkLru, err := lru.New(maxBlocks)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create lru cache for blocks: %w", err)
+	}
+
+	return &AllShardsReadBlockstore{
+		d:               d,
+		shardSelectF:    shardSelector,
+		blockstoreCache: bslru,
+		blockCache:      blkLru,
+	}, nil
+}
+
+func (ro *AllShardsReadBlockstore) Get(ctx context.Context, c cid.Cid) (b blocks.Block, finalErr error) {
+	logbs.Debugw("bitswap Get called", "cid", c)
```

Comment:
I suggest removing the word "bitswap" from all of these log lines as it may not necessarily be bitswap that calls uses this read-only blockstore

Reply:
This is done in the new `IndexBackedBlockstore ` type.